h2o-admissibleml/src/main/java/hex/Infogram/Infogram.java

```diff
@@ -0,0 +1,408 @@
+        if (model != null) {
+          keepFrameKeys(keep, _cmiRelKey);
+        }
+        Scope.exit(keep.toArray(new Key[keep.size()]));
+        model.update(_job._key);
+        model.unlock(_job);
+      }
+    }
+
+    /**
+     * Copy over info to model._output for _cmi_raw, _cmi, _topKFeatures,
+     * _all_predictor_names.  Derive _admissible for predictors if cmi >= cmi_threshold and 
+     * relevance >= relevance_threshold.  Derive _admissible_index as distance from point with cmi = 1 and 
+     * relevance = 1.  In addition, all arrays are sorted on _admissible_index.
+     */
+    private void copyCMIRelevance(InfogramModel.InfogramModelOutput modelOutput) {
+      modelOutput._cmi_raw = new double[_cmi.length];
+      System.arraycopy(_cmiRaw, 0, modelOutput._cmi_raw, 0, modelOutput._cmi_raw.length);
+      modelOutput._admissible_index = new double[_cmi.length];
+      modelOutput._admissible = new double[_cmi.length];
+      modelOutput._cmi = _cmi.clone();
+      modelOutput._topKFeatures = _topKPredictors.clone();
+      modelOutput._all_predictor_names = _topKPredictors.clone();
+      int numRows = _varImp.getRowDim();
+      String[] varRowHeaders = _varImp.getRowHeaders();
+      List<String> relNames = new ArrayList<String>(Arrays.asList(varRowHeaders));
+      modelOutput._relevance = new double[numRows];
+      double normalizedAdmissibleIndex = 1.0/Math.sqrt(2.0);
+      for (int index = 0; index < numRows; index++) { // extract predictor with varimp >= threshold
+        int newIndex = relNames.indexOf(modelOutput._all_predictor_names[index]);
+        modelOutput._relevance[index] = (double) _varImp.get(newIndex, 1);
+        double relDiff = modelOutput._relevance[index];
+        double cmiDiff = _cmi[index];
+        modelOutput._admissible_index[index] = normalizedAdmissibleIndex*Math.sqrt(relDiff*relDiff+cmiDiff*cmiDiff);
+        modelOutput._admissible[index] = (modelOutput._relevance[index] >= _parms._relevance_threshold
+                && modelOutput._cmi[index] >= _parms._cmi_threshold) ? 1 : 0;
+      }
+      int[] indices = IntStream.range(0, _cmi.length).toArray();
+      sort(indices, modelOutput._admissible_index, -1, -1);
+      modelOutput.sortCMIRel(indices);
+    }
+
+    /***
+     * Top level method to break down the infogram process into parts.
+     * 
+     * I have a question here for all of you:  Instead of generating the training frame and model builders for all
+     * the predictors, I break this down into several parts with each part generating _parms._nparallelism training
+     * frames and model builders.  For each part, after _parms._nparallelism models are built, I extract the entropy
+     * for each predictor.  Then, I move to the next part.  My question here is:  is this necessary?  I am afraid of
+     * the memory consumption of spinning up so many training frames and model builders.  If this is not an issue,
+     * please let me know.
```

Comment:
@maurever Yes, I do worry about running this with large dataset.  I have done some optimization in the sense that I only run _parms._nparallelism models at a time to hopefully avoid running out of memory.  It will be crucial to choose the correct number here to not blow out the memory and not to run incredible slow either.

Reply:
Ok, so maybe in the documentation, it should be mentioned how to set correctly `nparallelism` parameter to avoid out of memory. ;)